Information about indexes affected by a given query

pranavk opened the pull request:
This is the first cut for #1097 

I can move the common functions like isDisjoint to some other more common utility class but first I want to get an idea if this is going in the right direction. :)

Also, I didn't create a new class as was described in the issue since there was already this PlanUtil in planner::

malin1993ml commented on src/include/planner/plan_util.h:
```diff
@@ -149,5 +169,92 @@ inline void PlanUtil::GetTablesReferenced(const planner::AbstractPlan *plan,
+    } break;
+    default:
+      LOG_TRACE("Does not support finding affected indexes for query type: %d",
+                static_cast<int>(sql_stmt->GetType()));
```
I think you should add a switch case for `SELECT` queries, which is just going to return an empty set.

malin1993ml commented on src/include/planner/plan_util.h:
```diff
@@ -149,5 +169,92 @@ inline void PlanUtil::GetTablesReferenced(const planner::AbstractPlan *plan,
+      setA_it++;
+    else
+      setB_it++;
+  }
```
Put brackets for all if statements for style consistency.

malin1993ml commented on src/include/planner/plan_util.h:
```diff
@@ -149,5 +169,92 @@ inline void PlanUtil::GetTablesReferenced(const planner::AbstractPlan *plan,
+  while (setA_it != setA.end() && setB_it != setB.end() && disjoint) {
+    if (*setA_it == *setB_it)
+      disjoint = false;
+    else if (*setA_it < *setB_it)
```
It seems that you're assuming an ascending order for both sets, which may not be true. I think you need to sort the two sets before doing this.

pranavk replied:
Hi,

The std::set stores the element inside in an ascending order already by default; so, there's no need to sort explicitly here.

pranavk replied:
But actually, yeah. Let's sort them to be safe. Though the current implementation provide only sorted sets to this utility function, being a utility function, our users can give us all sorts of crazy input sets; so, better safe than sorry :)

malin1993ml commented:
@saatviks You should also look at this PR if you have time, because this is related to the stuff you're going to do later.

malin1993ml commented:
@pranavk These all look good! However, you need to add test cases to test whether your `GetAffectedIndexes` function can extract the affected indexes for the four types of queries correctly. To do that, you need to create a table, create several indexes, and use your function to analyze some queries. Actually as I'm thinking of this, [this test case ](https://github.com/cmu-db/peloton/blob/master/test/optimizer/old_optimizer_test.cpp) is probably a better reference for you. You can look at how it creates the database, table, and gets the statement from the SQL string.

pmenon commented on src/include/planner/plan_util.h:
```diff
@@ -44,6 +53,15 @@ class PlanUtil {
+  * @return set of affected index object ids
+  */
+  static const std::set<oid_t> GetAffectedIndexes(
+      catalog::CatalogCache &catalog_cache, parser::SQLStatement *sql_stmt);
```
Can `sql_stmt` be null? If not, it's probably a good idea to make it a reference. If the function doesn't modify it, even better to make it `const`.

pranavk replied:
Good point. I have made it a const reference and made appropriate changes required for it.

pmenon commented on src/include/planner/plan_util.h:
```diff
@@ -25,6 +25,15 @@
+
+namespace parser {
+class SQLStatement;
+}
```
Namespace closures for all these.

pmenon commented on src/include/catalog/catalog_cache.h:
```diff
@@ -18,6 +18,11 @@
+
+namespace planner {
+    class PlanUtil;
+}
```
1. Namespace closures.
2. Indentation.

pmenon commented on src/include/util/set_util.h:
```diff
@@ -0,0 +1,63 @@
+      } else {
+        setB_it++;
+      }
+    }
```
1. Rename to `SetUtil::IsDisjoint()`
2. Move implementation outside class definition as per [style guide](https://github.com/cmu-db/peloton/blob/master/script/coding_style.md).
3. Is the sortedness check required? `std::set` provides ordered iteration (ascending by default).

pranavk replied:
> Is the sortedness check required? std::set provides ordered iteration (ascending by default).

Yeah, true. I did the same thing before, i.e., didn't sort it. But changed the behavior after @malin1993ml 's comment above thinking that the input set could have been initialized by the user with a different comparator. But now when I think about it again, it's impossible since the input params to this function are only set<T> and not set<T, CmpFunc>. I'll remove the superfluous sorting operation again.

malin1993ml replied:
Sounds reasonable! I didn't know that `std::set` comes with default ascending order.

pmenon commented on src/include/util/set_util.h:
```diff
@@ -0,0 +1,46 @@
+  if (setA.empty() || setB.empty()) return disjoint;
+
+  typename std::set<T>::const_iterator setA_it = setA.begin();
+  typename std::set<T>::const_iterator setB_it = setB.begin();
```
You can use `auto` here to reduce visual noise.

pranavk replied:
I thought it doesn't work in C++11 but it does!! great! :) I will fix it.

pmenon commented on src/planner/plan_util.cpp:
```diff
@@ -0,0 +1,100 @@
+//
+// Identification: src/planner/plan_util.cpp
+//
+// Copyright (c) 2015-18, Carnegie Mellon University Database Group
```
Fix copyright.

pranavk replied:
Change what? Write it as 2018 instead of 18? We should also fix it in the formatter.py script so that it complains for mistakes like this.

pranavk replied:
And FWIW, the same pattern, eg: 2015-17 is being followed in many other files as well. See create_plan.cpp , eg:.

pmenon replied:
Correct, years should be full. In general, if you run the formatter script [here](https://github.com/cmu-db/peloton/blob/master/script/formatting/formatter.py) to add the header, it should do it correctly.

You're correct that other files don't adhere to the rules. This is something we're fixing incrementally rather than submitting a giant PR that breaks other ongoing PRs.

pranavk replied:
cool; pushed the new changes. Yeah, turns out I missed the `-u` flag to the script.

pranavk commented:
@pbollimp There's nothing in-progress about this PR. It's ready for further review or merge, if it looks good to someone.

malin1993ml commented:
@pranavk Thanks for updating the PR. However, we still need a test case like I commented before:
https://github.com/cmu-db/peloton/pull/1105#issuecomment-363864343

After that I think we're good to go.

pranavk commented:
@malin1993ml Ah, I commented on a wrong PR. 

And yes, thanks for reminding.

pranavk commented:
So, I tried to make use of helper functions in `TestingExecutorUtil` as mentioned in the test/README file; I tried using `CreateAndPopulateTable` (this is what is recommended in the README file). But this utility function only creates the table on the storage layer and doesn't insert table, column, or the index into their respective catalog objects. Since the new function `GetAffectedIndexes` relies on data being available in the `CatalogCache`, things doesn't work as expected inside `GetAffectedIndexes`.

The next thing  to try in my list is to use the `Catalog` object's methods such as AddTable, CreateTable, CreateIndex, etc. so that things really are available in the CatalogCache before I call `GetAffectedIndex`. But there are some more tests already that does this thing (cf. DropTests). My question is should I factor this db, table, index creation under a new set of `TestingExecutorUtil` or modify the existing `CreateAndPopulateTable` to add things to Catalog after creating these?

pranavk commented:
@malin1993ml I've added the test as requested :)

malin1993ml commented on src/include/parser/update_statement.h:
```diff
@@ -56,6 +56,10 @@ class UpdateStatement : public SQLStatement {
 
+  inline void TryBindDatabaseName(std::string default_database_name) {
+    table->TryBindDatabaseName(default_database_name);
+  }
```
We can't add a function that is only used in the test to the source code. We have to use something else.

malin1993ml commented on test/planner/plan_util_test.cpp:
```diff
@@ -0,0 +1,160 @@
+  // all indexes are affected
+  EXPECT_EQ(static_cast<int>(affected_indexes.size()), 2);
+  expected_oids = std::set<oid_t>({id_idx_oid, fname_idx_oid});
+  EXPECT_EQ(expected_oids, affected_indexes);
```
You also want to add a select case which will affect no index.

malin1993ml commented on test/planner/plan_util_test.cpp:
```diff
@@ -0,0 +1,160 @@
+  catalog->CreateDatabase(TEST_DB_NAME, txn);
+  auto db = catalog->GetDatabaseWithName(TEST_DB_NAME, txn);
+  // Insert a table first
+  auto id_column = catalog::Column(
```
We should change how we do this and creating indexes.

pranavk replied:
You mean using the approach in old_optimizer_test? I looked at both the approaches and this looked better to me (using methods instead of manually typing queries to create tables, indexes in). And I couldn't find any downsides of this approach.

malin1993ml replied:
Oh if you can get it work without adding the extra TryBindDatabaseName function then I think it's fine. Then like you said, it would would be better to factor this db, table, index creation under a new set of `TestingExecutorUtil`, and ideally update the DropTests to use this as well.

pranavk replied:
I have already done so. :)

malin1993ml replied:
Cool! Can you factor this out to TestingExecutorUtil as a new utility function as well? Then other tests can also use this. Then I think it’s only the compilation issue left.

malin1993ml commented on test/planner/plan_util_test.cpp:
```diff
@@ -0,0 +1,160 @@
+                       "test_fname_idx", false, IndexType::BWTREE, txn);
+  txn_manager.CommitTransaction(txn);
+
+  // dummy txn to get the catalog_cache object
```
The rest of the test looks good.

malin1993ml commented:
@pranavk oh also it seems that you need to update your PR with the master and pass the Travis test.

pranavk commented:
@malin1993ml I didn't update the PR since master, for me on F27, doesn't compile.

malin1993ml commented:
@pranavk I see. I'll check the compilation issue with others.

malin1993ml commented:
Hi @pranavk, so we fixed a compilation issue with Fedora 27. Please try to see whether you can compile it now. If so, there're three minors things left:
1. Update your PR with master
2. Fix the Travis build
3. Factor the db, table, index creation in the test under a new set of TestingExecutorUtil

pranavk commented:
@malin1993ml I updated the PR with master and fixed the Travis build.

Regarding factoring out the code, I was thinking it might be better to handle this in a separate PR since the  code copy & pasta was already there before the PR. It's not just the drop tests with which my code matches. There are several tests like constraint_tests, etc. See `git grep CreateDatabase test/` for more.

malin1993ml commented:
@pranavk Make sense. Then can you write an issue about what should we factor out later and where are they? In terms of the code, I think we're done here:)

Can you update your PR again? Github tells me that it's still out-of-date.

pmenon commented on test/planner/plan_util_test.cpp:
```diff
@@ -0,0 +1,170 @@
+      planner::PlanUtil::GetAffectedIndexes(txn->catalog_cache, *sql_stmt);
+
+  // id and first_name are affected
+  EXPECT_EQ(static_cast<int>(affected_indexes.size()), 2);
```
For these assertions/expectations, the expected value should be the first argument.

pmenon commented on test/planner/plan_util_test.cpp:
```diff
@@ -0,0 +1,170 @@
+  // no indexes are affected
+  EXPECT_EQ(static_cast<int>(affected_indexes.size()), 0);
+}
+}
```
Properly close the namespaces.

pmenon commented on test/util/set_util_test.cpp:
```diff
@@ -0,0 +1,46 @@
+//
+// set_util_test.cpp
+//
+// Identification: test/set_util_test.cpp
```
Path is wrong.

pmenon commented on test/planner/plan_util_test.cpp:
```diff
@@ -0,0 +1,170 @@
+  txn = txn_manager.BeginTransaction();
+  auto source_table = db->GetTableWithName("test_table");
+  oid_t col_id = source_table->GetSchema()->GetColumnID(id_column.column_name);
+  LOG_INFO("id: %d", static_cast<int>(col_id));
```
Just curious, are these logs needed in the test? What information do they convey?

pranavk replied:
Ah, these are the debug logs leftover from my testing. Not needed.

pranavk commented:
@pmenon Cleaned up
